pkgs/applications/networking/instant-messengers/signal-desktop/default.nix

```
@@ -1,6 +1,4 @@
-{ hostPlatform, callPackage }: {
-  signal-desktop = if hostPlatform.system == "aarch64-linux"
-    then callPackage ./signal-desktop-aarch64.nix { }
-    else callPackage ./signal-desktop.nix { };
-  signal-desktop-beta = callPackage ./signal-desktop-beta.nix{ };
+{ targetPlatform, callPackage }: {
+  signal-desktop = callPackage ./signal-desktop-${targetPlatform.system}.nix { };
```

Mhm. This will cause eval errors in case we have a new unsupported platform.
This needs to be fixed.

---

Maybe builtins.pathExists with a fallback to x86_64-linux? But otherwise the previous if statement was also good enough, easier to read than some other complex logic.

---

Perhaps we could do something like this?
```suggestion
  signal-desktop =
    let
      default = callPackage ./signal-desktop-x86_64-linux.nix { };
      supportedPlatforms = default.meta.platforms;
      inherit (targetPlatform) system;
    in
    if builtins.elem system supportedPlatforms
    then callPackage ./signal-desktop-${system}.nix { }
    else default;
```
Or we could use `builtins.pathExists` like this
```suggestion
  signal-desktop =
    let
      path = ./signal-desktop-${targetPlatform.system}.nix;
      default = ./signal-desktop-x86_64-linux.nix;
    in
    if builtins.pathExists path
    then callPackage path { }
    else callPackage default { };
```
I think I prefer the latter. Evaluating `meta` in the first one would work but it's harder to follow.

---

@Mic92 Wouldn't it be better to just `throw` and say your platform is not supported? Since there isn't any guarantee the `x86_64-linux` will work properly on a new platform either?

---

I think a throw statement is not quite same as saying that the platform is not supported in `meta.platforms` semantically speaking.

---

We already have quite a few eval errors here and we should not keep adding to it: https://hydra.nixos.org/jobset/nixpkgs/trunk#tabs-errors

---

> I think a throw statement is not quite same as saying that the platform is not supported in meta.platforms semantically speaking.

You're right, although what now? Do we default to building `x86_64-linux`? Wouldn't it be better to mark anything that isn't in `meta.platforms` as broken so we can entirely skip making a derivation and save resources? 

Also, if we were to default to building for `x86_64-linux`, wouldn't it be an issue if we could build it on another platform but not run it?

---

According to @wegank, `targePlatform` is only required when building a compiler https://github.com/NixOS/nixpkgs/pull/310053#discussion_r1594205907, see https://nix.dev/tutorials/cross-compilation.html#platforms.

---

> and run the final executable on the target platform.

If `hostPlatform` were used and you were to cross compile architectures, wouldn't you end up with a binary that's incompatible if `hostPlatform` were used?

---

The cross-compiled program, in that example the compiler, still runs on the host platform (and produces executables that run on the target platform).